File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/passes/RemoveUnreachableBlocks.java
package ro.pub.cs.diploma.passes;

import com.intellij.openapi.util.Ref;
import org.jetbrains.annotations.NotNull;
import ro.pub.cs.diploma.ir.Block;

import java.util.*;
import java.util.stream.Collectors;

public class RemoveUnreachableBlocks implements Pass<List<Block>, List<Block>, List<Block>> {
  private RemoveUnreachableBlocks() {

  }

  @NotNull
  public static RemoveUnreachableBlocks getInstance() {
    return new RemoveUnreachableBlocks();
  }

  @Override
  @NotNull
  public List<Block> collect(@NotNull final List<Block> blocks) {
    return blocks;
  }

  @NotNull
  private static Set<Block> getReachableBlocks(@NotNull final Block root) {
    final Set<Block> reachableBlocks = new HashSet<>();
    reachableBlocks.add(root);

    final Queue<Block> queue = new ArrayDeque<>();
    queue.add(root);

    while (!queue.isEmpty()) {
      final Block current = queue.remove();
      for (Ref<Block> ref : current.getOutBlocks()) {
        final Block child = ref.get();
        if (!reachableBlocks.contains(child)) {
          reachableBlocks.add(child);
          queue.add(child);
        }
      }
    }

    return reachableBlocks;
  }

  @Override
  @NotNull
  public List<Block> transform(@NotNull final List<Block> blocks) {
    final Set<Block> reachableBlocks = getReachableBlocks(blocks.get(0));
    return blocks.stream().filter(reachableBlocks::contains).collect(Collectors.toList());
  }
}

File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/passes/RemoveUnreachableBlocks.kt
package ro.pub.cs.diploma.passes

import ro.pub.cs.diploma.ir.Block
import java.util.*

class RemoveUnreachableBlocks : Pass<List<Block>, List<Block>, List<Block>> {

  override fun collect(blocks: List<Block>): List<Block> = blocks

  override fun transform(blocks: List<Block>): List<Block> {
    val reachableBlocks = getReachableBlocks(blocks[0])
    return blocks.filter { reachableBlocks.contains(it) }.toList()
  }

  private fun getReachableBlocks(root: Block): Set<Block> {
    val reachableBlocks = HashSet<Block>()
    reachableBlocks.add(root)

    val queue = ArrayDeque<Block>()
    queue.add(root)

    while (!queue.isEmpty()) {
      val current = queue.remove()
      for (ref in current.outBlocks) {
        val child = ref.get()
        if (!reachableBlocks.contains(child)) {
          reachableBlocks.add(child)
          queue.add(child)
        }
      }
    }

    return reachableBlocks
  }
}
